### PR TITLE
Pagecontainer classes

### PR DESCRIPTION
**Pull request by arschmitz**
This incorporates and updates #7947 this now uses the classes option to add `.ui-pagecontainer` and uses that as the canonical way to find a page container

### CODE REVIEW

File: demos/backbone-requirejs/index.php [hunk 1]
```diff
@@ -115,7 +115,7 @@
       } );
 </code>
 </pre>
-<p>Next, inside of the Backbone.js Router class object, we can respond to haschange events and manually call the jQuery Mobile <code>changePage()</code> method.  Below is a small snippet of <strong>mobileRouter.js</strong>.</p>
+<p>Next, inside of the Backbone.js Router class object, we can respond to haschange events and manually call the <code>pagecontainer( "change" )</code> method.  Below is a small snippet of <strong>mobileRouter.js</strong>.</p>
```
Comment (gabrielschulhof) [inline demos/backbone-requirejs/index.php]: ```.pagcontainer( "change", ... )``` (notice the period in front of "pagecontainer") the line above has that period.

File: demos/backbone-requirejs/index.php [hunk 2]
```diff
@@ -133,7 +133,10 @@
         home: function() {
 
             // Programatically changes to the categories page
-            $.mobile.changePage( "#categories" , { reverse: false, changeHash: false } );
+            $( ".ui-pagecontainer" ).pagecontainer( "change", "#categories" , {
```
Comment (gabrielschulhof) [inline demos/backbone-requirejs/index.php]: We need to use ```.closest( ".ui-pagecontainer" )``` ...

File: demos/backbone-requirejs/js/routers/mobileRouter.js [hunk 3]
```diff
@@ -16,6 +16,8 @@ define([
         // The Router constructor
         initialize: function() {
 
+            this.pagecontainer = $( ".ui-pagecontainer" ).pagecontainer( "instance" );
```
Comment (gabrielschulhof) [inline demos/backbone-requirejs/js/routers/mobileRouter.js]: We need to use ```.closest( ".ui-pagecontainer" )``` ...

File: css/structure/jquery.mobile.collapsible.css [hunk 4]
```diff
@@ -97,3 +97,6 @@
 .ui-collapsible-set .ui-collapsible.ui-first-child {
 	margin-top: 0;
 }
+.ui-collapsible-icon-right {
+	float: right;
```
Comment (gabrielschulhof) [inline css/structure/jquery.mobile.collapsible.css]: This belongs in a different PR.

File: demos/collapsible/index.php [hunk 5]
```diff
@@ -111,11 +111,30 @@
...
 
-            <div data-role="collapsible" data-iconpos="right">
-                <h3>I'm a header</h3>
-                <p><code>data-iconpos="right"</code></p>
+            <div data-demo-html="true">
+                <div data-role="collapsible" data-iconpos="right">
+                    <h3>I'm a header</h3>
+                    <p><code>data-iconpos="right"</code></p>
+                </div>
+            </div>
```
Comment (gabrielschulhof) [inline demos/collapsible/index.php]: This belongs into a collapsible PR.

File: demos/collapsible/index.php [hunk 6]
```diff
@@ -225,8 +244,9 @@
 				<div data-demo-html="true">
 					<div data-role="collapsible" data-enhanced="true" class="ui-collapsible ui-collapsible-inset ui-corner-all ui-collapsible-collapsed">
 						<h4 class="ui-collapsible-heading ui-collapsible-heading-collapsed">
-							<a href="#" class="ui-collapsible-heading-toggle ui-button ui-icon-plus">
-							Heading
+							<a href="#" class="ui-collapsible-heading-toggle ui-button">
+                            <span class="ui-icon ui-icon-plus"></span>
```
Comment (gabrielschulhof) [inline demos/collapsible/index.php]: This belongs into a collapsible PR.

File: demos/old-faq-pages/index.php [hunk 7]
```diff
@@ -9,7 +9,7 @@
 <li data-role="list-divider">Pages</li>
 <li data-section="Questions &amp; Answers" data-filtertext="head load scripts styles pageinit onload pagebeforecreate ajax nav pages"><a href="faq/scripts-and-styles-not-loading.php">Why aren't my scripts and styles loading?</a></li>
 <li data-section="Questions &amp; Answers" data-filtertext="document.ready onload paginit DOM ready on ready scripts "><a href="faq/dom-ready-not-working.php">Why isn't DOM ready working for my scripts?</a></li>
-<li data-section="Questions &amp; Answers" data-filtertext="ajax nav load page mobile.loadpage"><a href="faq/how-do-i-load-a-page.php">How do I load a page?</a></li>
+<li data-section="Questions &amp; Answers" data-filtertext="ajax nav load page pagecontainer load"><a href="faq/how-do-i-load-a-page.php">How do I load a page?</a></li>
```
Comment (gabrielschulhof) [inline demos/old-faq-pages/index.php]: "load" appears twice, so that makes the second one redundant, AFAIK.

File: js/navigation.js [hunk 8]
```diff
@@ -425,13 +394,16 @@ $.mobile._registerInternalEvents = function() {
...
+					role: $link.attr( "data-" + $.mobile.ns + "rel" ),
+					prefetch: true
+				});
 			}
 		} );
 	} );
 
 	// TODO ensure that the navigate binding in the content widget happens at the right time
-	$.mobile.pageContainer.pagecontainer();
+	//$.mobile.pageContainer.pagecontainer();
```
Comment (gabrielschulhof) [inline js/navigation.js]: We shouldn't be commented out code. Just remove it.

File: js/widgets/collapsible.js [hunk 9]
```diff
@@ -30,13 +30,7 @@
 	}
```
Comment (gabrielschulhof) [inline js/widgets/collapsible.js]: This belongs in a collapsible PR.

File: js/widgets/pagecontainer.js [hunk 10]
```diff
@@ -79,10 +103,10 @@ $.widget( "mobile.pagecontainer", {
 
 	_setOptions: function( options ) {
 		if ( options.theme !== undefined && options.theme !== "none" ) {
-			this.element.removeClass( "ui-overlay-" + this.options.theme )
-				.addClass( "ui-overlay-" + options.theme );
+			this._removeClass( null, "ui-overlay-" + this.options.theme )
```
Comment (gabrielschulhof) [inline js/widgets/pagecontainer.js]: You should be able to do this with the theme extension.

File: js/widgets/pagecontainer.js [hunk 11]
```diff
@@ -1186,6 +1206,19 @@ $.widget( "mobile.pagecontainer", {
...
 		return closestBase || $.mobile.path.documentBase.hrefNoHash;
+	},
+
+	_destroy: function() {
+		var myIndex;
+
+		if ( $.mobile.pagecontainers ) {
+			myIndex = $.inArray( this.element, $.mobile.pagecontainers );
+			if ( myIndex >= 0 ) {
+				$.mobile.pagecontainers.splice( myIndex, 1 );
```
Comment (gabrielschulhof) [inline js/widgets/pagecontainer.js]: What if I'm the active pagecontainer? ```.active``` should then be ... unset?

File: tests/integration/checkboxradio/checkboxradio_core.js [hunk 12]
```diff
@@ -0,0 +1,150 @@
+/*
```
Comment (gabrielschulhof) [inline tests/integration/checkboxradio/checkboxradio_core.js]: This doesn't seem related to pagecontainer.

File: tests/integration/collapsible/collapsible_core.js [hunk 13]
```diff
@@ -7,18 +7,18 @@
 function testExpandCollapse( selector ) {
```
Comment (gabrielschulhof) [inline tests/integration/collapsible/collapsible_core.js]: This belongs in the collapsible PR.

File: tests/integration/degrade-inputs/degradeInputs.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
+/*
```
Comment (gabrielschulhof) [inline tests/integration/degrade-inputs/degradeInputs.js]: This also seems like it belongs in a different PR.

File: tests/integration/dialog-extension/dialog_events.js [hunk 15]
```diff
@@ -93,11 +93,11 @@ QUnit.asyncTest( "Clicking dialog 'Close' button renders it unclickable", functi
 
 	$.testHelper.pageSequence( [
 		function() {
-			$.mobile.changePage( $( "#mypage" ) );
+			$( ".ui-pagecontainer" ).pagecontainer( "change", $( "#mypage" ) );
```
Comment (gabrielschulhof) [inline tests/integration/dialog-extension/dialog_events.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/dialog-extension/dialog_no_hash.js [hunk 16]
```diff
@@ -15,7 +15,7 @@ asyncTest( "dialog opens and closes correctly when hash handling is off", functi
 
 	$.testHelper.pageSequence( [
 		function() {
-			$.mobile.changePage( $( "#mypage" ) );
+			$( ".ui-pagecontainer" ).pagecontainer( "change", $( "#mypage" ) );
```
Comment (gabrielschulhof) [inline tests/integration/dialog-extension/dialog_no_hash.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/dialog/dialog_events.js [hunk 17]
```diff
@@ -59,7 +59,7 @@ QUnit.asyncTest( "dialog element with no theming", function( assert ) {
 
 	$.testHelper.pageSequence( [
 		function() {
-			$.mobile.changePage( $( "#mypage" ) );
+			$( ".ui-pagecontainer" ).pagecontainer( "change", $( "#mypage" ) );
```
Comment (gabrielschulhof) [inline tests/integration/dialog/dialog_events.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/dialog/dialog_no_hash.js [hunk 18]
```diff
@@ -15,7 +15,7 @@ asyncTest( "dialog opens and closes correctly when hash handling is off", functi
 
 	$.testHelper.pageSequence( [
 		function() {
-			$.mobile.changePage( $( "#mypage" ) );
+			$( ".ui-pagecontainer" ).pagecontainer( "change", $( "#mypage" ) );
```
Comment (gabrielschulhof) [inline tests/integration/dialog/dialog_no_hash.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/fixed-toolbar/fixedToolbar.js [hunk 19]
```diff
@@ -163,9 +163,9 @@ asyncTest( "The toggle method is working properly", function() {
 asyncTest( "Fullscreen toolbars add classes to page", function() {
 	expect( 2 );
 
-	$.testHelper.sequence( [
-		function() {
-			$.mobile.changePage( "#fullscreen-test-a" );
+	$.testHelper.sequence([
```
Comment (gabrielschulhof) [inline tests/integration/fixed-toolbar/fixedToolbar.js]: There's a whitespace regression here.

File: tests/integration/fixed-toolbar/fixedToolbar.js [hunk 20]
```diff
@@ -190,10 +190,11 @@ asyncTest( "The persistent headers and footers are working properly", function()
 		nextpagefooter = $( "#persist-test-b .ui-footer-fixed" );
 
 
-	$.testHelper.pageSequence( [
-		function() {
+
+	$.testHelper.pageSequence([
```
Comment (gabrielschulhof) [inline tests/integration/fixed-toolbar/fixedToolbar.js]: There's a whitespace regression here.

File: tests/integration/listview/listview_core.js [hunk 21]
```diff
@@ -25,7 +25,7 @@ asyncTest( "The page should be enhanced correctly", function() {
 	expect( 3 );
 	$.testHelper.pageSequence( [
 		function() {
-			$.mobile.changePage( "#basic-linked-test" );
+			$( ".ui-pagecontainer" ).pagecontainer( "change", "#basic-linked-test" );
```
Comment (gabrielschulhof) [inline tests/integration/listview/listview_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/navigation/navigation_base.js [hunk 22]
```diff
@@ -90,8 +90,8 @@ asyncTest( "can navigate between internal and external pages", function( assert
 				report: "navigate from a page in a non-base directory to an internal page"
 			} );
 
-			// Try calling changePage() directly with a relative path.
-			$.mobile.changePage( "base-page-1.html" );
+			// Try calling change() directly with a relative path.
+			$( ".ui-pagecontainer" ).pagecontainer( "change", "base-page-1.html" );
```
Comment (gabrielschulhof) [inline tests/integration/navigation/navigation_base.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/navigation/navigation_core.js [hunk 23]
```diff
@@ -127,7 +127,7 @@ QUnit.asyncTest( "window.history.back() from external to internal page", functio
 QUnit.asyncTest( "external empty page does not result in any contents", function( assert ) {
 	$.testHelper.pageSequence( [
 		function() {
-			$.mobile.changePage( "blank.html" );
+			$( ".ui-pagecontainer" ).pagecontainer( "change", "blank.html" );
```
Comment (gabrielschulhof) [inline tests/integration/navigation/navigation_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/navigation/sequence/sequence_core.js [hunk 24]
```diff
@@ -51,7 +51,7 @@ asyncTest( "Navigating to an internal page", function() {
 			$( "#openInternalPage" ).click();
 		},
 		{
-			pagecontainerchange: { src: $.mobile.pageContainer, event: "pagecontainerchange" + eventNs + "1" }
+			pagecontainerchange: { src: $( ".ui-pagecontainer" ), event: "pagecontainerchange" + eventNs + "1" }
```
Comment (gabrielschulhof) [inline tests/integration/navigation/sequence/sequence_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/pagecontainer/page_load_failure_core.js [hunk 25]
```diff
@@ -15,13 +15,13 @@ $( document ).on( "pagecontainerloadfailed", function( event ) {
 				$( "#go-to-other-page" ).click();
 			},
 			function() {
-				deepEqual( $.mobile.pageContainer.pagecontainer( "getActivePage" )
+				deepEqual( $( ".ui-pagecontainer" ).pagecontainer( "getActivePage" )
```
Comment (gabrielschulhof) [inline tests/integration/pagecontainer/page_load_failure_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/popup/popup_core.js [hunk 26]
```diff
@@ -158,7 +158,7 @@ asyncTest( "Popup does not go back in history twice when opening on separate pag
 			$( "#go-to-another-page" ).click();
 		},
 		{
-			pagechange: { src: $.mobile.pageContainer, event: "pagechange" + eventNs + "1" }
+			pagechange: { src: $( ".ui-pagecontainer" ), event: "pagechange" + eventNs + "1" }
```
Comment (gabrielschulhof) [inline tests/integration/popup/popup_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/select/button_focus_core.js [hunk 27]
```diff
@@ -15,7 +15,7 @@ function defineTest( testName, clickAction, expectChange ) {
 				}
 			},
 			function( result ) {
-				var activePage = $.mobile.pageContainer.pagecontainer( "getActivePage" );
+				var activePage = $( ".ui-pagecontainer" ).pagecontainer( "getActivePage" );
```
Comment (gabrielschulhof) [inline tests/integration/select/button_focus_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/select/select_cached.js [hunk 28]
```diff
@@ -22,7 +22,7 @@ QUnit.asyncTest( "dialog sized select should alter the value of its parent selec
 			resetHash,
 
 			function() {
-				$.mobile.changePage( "cached.html" );
+				$( ".ui-pagecontainer" ).pagecontainer( "change", "cached.html" );
```
Comment (gabrielschulhof) [inline tests/integration/select/select_cached.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/select/select_core.js [hunk 29]
```diff
@@ -431,7 +431,7 @@ QUnit.asyncTest( "dialog size select title should match the label when changed a
 
 			{
 				pagechange: {
-					src: $.mobile.pageContainer,
+					src: $( ".ui-pagecontainer" ),
```
Comment (gabrielschulhof) [inline tests/integration/select/select_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/integration/transitions/transitions_core.js [hunk 30]
```diff
@@ -97,12 +97,12 @@ NOTES:
...
 To is required, so each async function must call start() onToComplete, not onFromComplete.
 */
-asyncTest( "changePage applies perspective class to mobile viewport for flip", function() {
+asyncTest( "change() applies perspective class to mobile viewport for flip", function() {
 	expect( 1 );
 
 	$.testHelper.pageSequence( [
 		function() {
-			$.mobile.changePage( "#foo" );
+			$( ".ui-pagecontainer" ).pagecontainer( "change", "#foo" );
```
Comment (gabrielschulhof) [inline tests/integration/transitions/transitions_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/jquery.testHelper.js [hunk 31]
```diff
@@ -227,7 +227,7 @@ $.testHelper = {
 			}, 10000 );
 
 			// bind the recursive call to the event
-			( self.eventTarget || $.mobile.pageContainer ).one( event, function( event, data ) {
+			( self.eventTarget || $( ".ui-pagecontainer" ) ).one( event, function( event, data ) {
```
Comment (gabrielschulhof) [inline tests/jquery.testHelper.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/unit/event/event_core.js [hunk 32]
```diff
@@ -25,12 +24,9 @@ QUnit.module( libName, {
 		// the collections existence in non touch enabled test browsers
 		$.Event.prototype.touches = [ { pageX: 1, pageY: 1 } ];
 
-		$.mobile.pageContainer = originalPageContainer || $( "body" );
-
-		$( $.mobile.pageContainer ).unbind( "throttledresize" );
+		$( ".ui-pagecontainer" ).unbind( "throttledresize" );
```
Comment (gabrielschulhof) [inline tests/unit/event/event_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/unit/init/init_core.js [hunk 33]
```diff
@@ -111,7 +111,7 @@ require( [
 				function() {
 					var firstPage = findFirstPage();
 
-					assert.deepEqual( $.mobile.pageContainer[ 0 ], firstPage.parent()[ 0 ] );
+					assert.deepEqual( $( ".ui-pagecontainer" )[ 0 ], firstPage.parent()[ 0 ] );
```
Comment (gabrielschulhof) [inline tests/unit/init/init_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/unit/page/page_core.js [hunk 34]
```diff
@@ -84,11 +84,11 @@ asyncTest( "page container is updated to page theme at pagebeforeshow", function
 
 	var pageTheme = "ui-overlay-" + $.mobile.activePage.page( "option", "theme" );
 
-	$.mobile.pageContainer.removeClass( pageTheme );
+	$( ".ui-pagecontainer" ).removeClass( pageTheme );
```
Comment (gabrielschulhof) [inline tests/unit/page/page_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/unit/panel/panel_core.js [hunk 35]
```diff
@@ -472,14 +472,14 @@ QUnit.asyncTest( "overlay panel should not call getWrapper", function( assert )
 		},
 		function() {
 			assert.deepEqual( count, 0, "getWrapper not called on close" );
-			$.mobile.changePage( "#page2" );
+			$( ".ui-pagecontainer" ).pagecontainer( "change", "#page2" );
```
Comment (gabrielschulhof) [inline tests/unit/panel/panel_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: tests/unit/table/table_core.js [hunk 36]
```diff
@@ -15,7 +15,7 @@ QUnit.module( "Basic Table", {
 				QUnit.start();
 			} );
 
-			$.mobile.changePage( hash );
+			$( ".ui-pagecontainer" ).pagecontainer( "change", hash );
```
Comment (gabrielschulhof) [inline tests/unit/table/table_core.js]: I guess this is OK for testing purposes since we control the content and we know there is only one pagecontainer. Still, maybe $.mobile.pagecontainers.active is in order?

File: demos/backbone-requirejs/index.php (hunk 2 above)
Comment (arschmitz) [inline demos/backbone-requirejs/index.php]: We can't use `closesest` `closeset` to what?

File: demos/backbone-requirejs/js/routers/mobileRouter.js (hunk 3 above)
Comment (arschmitz) [inline demos/backbone-requirejs/js/routers/mobileRouter.js]: We can't use `closest` closest to what?

File: js/widgets/pagecontainer.js (hunk 11 above)
Comment (arschmitz) [inline js/widgets/pagecontainer.js]: either unset or be changed to first one which still exists if others still exist

File: tests/integration/checkboxradio/checkboxradio_core.js (hunk 12 above)
Comment (arschmitz) [inline tests/integration/checkboxradio/checkboxradio_core.js]: Odd i didn't write this and have no idea where it came from?  These should not even exist these look like the old tests.

Comment (arschmitz) [inline tests/integration/checkboxradio/checkboxradio_core.js]: OHHHHHH!!!!! your PR i took over was before this was deleted so it added it back when merging!

File: js/widgets/pagecontainer.js (hunk 11 above)
Comment (arschmitz) [inline js/widgets/pagecontainer.js]: im thinking something like 

```
if ( $.mobile.pagecontainers.length ) {
	$.mobile.pagecontainers.active = $.mobile.pagecontainers[ 0 ];
} else {
	$.mobile.pagecontainers.active = undefined;
}
```

File: tests/integration/degrade-inputs/degradeInputs.js (hunk 14 above)
Comment (arschmitz) [inline tests/integration/degrade-inputs/degradeInputs.js]: same as checkboxradio above

File: tests/unit/event/event_core.js (hunk 32 above)
Comment (arschmitz) [inline tests/unit/event/event_core.js]: this should actually be `$( "body" )` `pagecontainer` is not even included in this test this is only passing because unbind will not throw on an empty collection

**Comment by arschmitz** [comment]
fixed and merged in 1.5-dev
